nixos/modules/services/cluster/patroni/default.nix

```diff
@@ -0,0 +1,249 @@
+        for possible values.
+        Secrets should be passed in by using the <literal>environment</literal> option.
+      '';
+    };
+
+    environment = mkOption {
+      type = with types; attrsOf (nullOr (oneOf [ str path package ]));
+      default = { };
+      example = {
+        PATRONI_REPLICATION_PASSWORD = "/secret/file";
+        PATRONI_SUPERUSER_PASSWORD = "/secret/file";
+      };
+      description = "Environment variables made available to Patroni, useful for providing secrets from files.";
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    services.patroni.settings = {
+      scope = cfg.scope;
+      name = cfg.name;
+      namespace = cfg.namespace;
+
+      restapi = {
+        listen = "${cfg.nodeIp}:${toString(cfg.restApiPort)}";
+        connect_address = "${cfg.nodeIp}:${toString(cfg.restApiPort)}";
+      };
+
+      raft = {
+        data_dir = "${cfg.dataDir}/raft";
+        self_addr = "${cfg.nodeIp}:5010";
+        partner_addrs = map (ip: ip + ":5010") cfg.otherNodesIps;
+      };
+
+      postgresql = {
+        listen = "${cfg.nodeIp}:${toString(cfg.postgresqlPort)}";
+        connect_address = "${cfg.nodeIp}:${toString(cfg.postgresqlPort)}";
+        data_dir = cfg.postgresqlDataDir;
+        bin_dir = "${postgresql}/bin";
+        pgpass = "${cfg.dataDir}/pgpass";
+      };
+      watchdog = mkIf cfg.watchdog {
+        mode = "required";
+        device = "/dev/watchdog";
+        safety_margin = 5;
+      };
+    };
+
+    users.users.postgres =
+      {
+        name = "postgres";
+        uid = config.ids.uids.postgres;
+        group = "postgres";
+        description = "PostgreSQL server user";
+        home = "${cfg.postgresqlDataDir}";
+        useDefaultShell = true;
+      };
+
+    users.groups.postgres.gid = config.ids.gids.postgres;
+
+    systemd.services = {
+      patroni = {
+        description = "Runners to orchestrate a high-availability PostgreSQL";
+
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+
+        environment = cfg.environment;
+
+        serviceConfig = mkMerge [{
+          User = "postgres";
+          Group = "postgres";
+          Type = "simple";
+          Restart = "on-failure";
+          TimeoutSec = 30;
+          ExecStart = "${patroni}/bin/patroni ${configFile}";
+          ExecReload = "${pkgs.coreutils}/bin/kill -s HUP $MAINPID";
+          KillMode = "process";
+        }
+          (mkIf (cfg.postgresqlDataDir == "/var/lib/postgresql/${postgresql.psqlSchema}" && cfg.dataDir == "/var/lib/patroni") {
+            StateDirectory = "patroni patroni/raft postgresql postgresql/${postgresql.psqlSchema}";
+            StateDirectoryMode = "0750";
+          })
+          (mkIf (cfg.watchdog) {
+            ExecStartPre = "+" + pkgs.writeShellScript "configure-watchdog.sh" ''
+              ${pkgs.kmod}/bin/modprobe softdog
+              ${pkgs.coreutils}/bin/chown postgres /dev/watchdog
```

Comment:
this looks extremely sketchy. what if a hardware watchdog exists? what if other services also want to use it?

Reply:
I wanted to provide this for convenience as I suppose this is the most common setup.
I can either remove the `watchdog` setting or rename it to `softwareWatchdog` and be explicit in the doc about what it does.

Reply:
not sure what the best course of action would be here. maybe split the option in two, one for *using* the watchdog and for setting up the software watchdog device. a note on how to configure a pre-existing watchdog for use with this service could also be helpful.

Reply:
I dont know how you would configure a pre-existing watchdog, Patroni [documentation](https://patroni.readthedocs.io/en/latest/watchdog.html#setting-up-software-watchdog-on-linux) doesn't explain that.
The documentation also states that *"watchdogs are only supported using Linux watchdog device interface"* so a single option to either use it or not seems fine.
It is also the first time I'm exposed to the concept of a watchdog so I'm not really qualified to give advice here.

Reply:
what's in the PR now will work fine. totally missed that `settings` itself references the watchdog, so using a hardware watchdog could just add those settings and leave `softwareWatchdog` unset.